### ISSUE TITLE
Scrape ads.txt, app-ads.txt and sellers.json

### ISSUE THREAD

**Issue by streitlua** [hestiaAI/ad-radar#21]
[ads.txt](https://iabtechlab.com/wp-content/uploads/2021/03/ads.txt-1.0.3.pdf) is a mechanism allowing publishers (a.k.a. website or content owners) to specify who are the parties authorized to sell their inventory (ad spaces/impressions).
Each publisher subdomain can have an `ads.txt` file, which is just a text file having a list of entries with 3 to 4 fields, like this one: `greenadexchange.com, XF7342, DIRECT, 5jyxf8k54`
Here's a description of the different fields:
1. (mandatory) specifies the domain name of an advertisement system
2. (mandatory) specifies the number of the account that the publisher uses on the advertisement system of field 1
3. (mandatory) specifies the relationship between the publisher and the advertisement system:
   a. DIRECT: there are no intermediaries, and there is likely a contract between the publisher and the advertisement system
   b. RESELLER: the publisher has authorized some third party to control its account (field 2) and to resell its ad space
4. (optional) certifies the advertisement system with some certificate authority

There is a similar mechanism called [app-ads.txt](https://iabtechlab.com/wp-content/uploads/2019/03/app-ads.txt-v1.0-final-.pdf) that allows an app owner to specify the parties authorized to sell their inventory.
Basically, the app page on the store that distributes it (e.g. Google Play) points to a domain that contains an `app-ads.txt` file with the same format as `ads.txt`.
So for

**Comment by loleg** [comment]
Some quick feedback, without knowing too much about this system, or even industry terminology like "inventory" or "ad exchange". This issue is a collection of ideas - to better define the work, I would break it down into issues with specific tasks as checklists. You could also publish this as concept document, and refer to sections in it inside of the issues. Otherwise the description is clear and readable. With some well defined tasks, I could see this type of project being easily distributed among multiple developers or even crowdsourced in its execution. 

The idea makes it seem like the scraping would need to be done simultaneously. However, more pragmatically you could aggregate the data from a set of publishers, then a set of advertisers, then cross-reference them after the fact. In general, scraping should avoid unnecessary repeat visits by caching the data collected, and refreshing as often as needed (which in this case is probably not very often).

It's interesting that you use topology to describe an idea. The web is a graph of nodes and relationships, I could see how you might use graph-based databases here, like neo4j ([medium post](https://medium.com/neo4j/graph-visualization-of-panama-papers-data-in-neo4j-9c08ca17039c), [moma example](https://github.com/mapmeld/graph)) or tigergraph ([tor example](https://www.tigergraph.com/blog/scraping-web-data-into-tigergraph-with-tor/)). So it might conversely to what I wrote above be more effective to not distinguish too much from the publishers and advertisers. Just create nodes and identify what files they serve to classify their data appropriately. You can disaggregate after the fact.

I wouldn't say that ads.txt is a mechanism: it's at best a standard. It would be good to think about not just the need to scrape the data, but also to validate it and help improve it as a service to the community. At least, this is what a good standards body like the W3C would do. Speaking of whom, I believe this is a topic of discussion in the [Credible Web Community](https://www.w3.org/community/credibility/).

Personally I like the initiative here of using open web data to raise transparency and accountability. Just try to do it in a well organized way and keep a wider perspective to avoid "shooting in the foot".

**Comment by mvidonne** [comment]
@streitlua  something like http://corrupt.marketing/?

**Comment by pdehaye** [comment]
Yes but this is not open data.

**Comment by foucault-dumas** [comment]
> @streitlua check the statistics for the top 400 French speaking websites done by the CNIL https://linc.cnil.fr/webpub-adstxt-sellersjson/ads_study.html @fquellec for the dataviz
> 
> https://github.com/LINCnil/Ads.txt-et-Sellers.json
> 
> under open licence 2.0

See also https://twitter.com/braedon/status/1468746918851272704?s=20

**Comment by foucault-dumas** [comment]
Another interesting tool: https://sellers.guide/

**Comment by mvidonne** [comment]
@ffsinger  is there a way to save properly all the info https://sellers.guide/domain/wikistrike.com

**Comment by ffsinger** [comment]
> @ffsinger is there a way to save properly all the info https://sellers.guide/domain/wikistrike.com

I haven't been following the discussions on these issues. What would you like to save ? The analysis result for a specific domain ? For a list of domains ? Which results specifically ? In a computer-readable or human-readable format ? For what purpose ?

Depending on the answer to these questions, we could build a (more or less complex) scraper.

**Comment by foucault-dumas** [comment]
sellers.guide just added a ads.txt cleaner on their page we just saw an interesting webinar by them with @mvidonne. She summarized it [here](https://drive.infomaniak.com/app/office/193995/92353).

I don't know how to answer @ffsinger's question, the idea is to use ads.txt and sellers.json to add live knowledge on the knowledge harvested on adds by adradar. For example display (and explain) the links between the intermediary identify as the winning bidder for an add and other actors of the adtech ecosystem. Can ads.txt and sellers.json (and sellers.guide) be used to do so?

**Comment by pdehaye** [comment]
Looks like it was an interesting seminar! 

I am thinking that this needs a strategy session involving @mvidonne, @foucault-dumas and myself, before involving developers more? Along the lines of "AdRadar needs to evolve towards helping find out during regular web browsing interesting situations for which it is worth investigating further the data angle"

**Comment by foucault-dumas** [comment]
> Looks like it was an interesting seminar!
> 
> I am thinking that this needs a strategy session involving @mvidonne, @foucault-dumas and myself, before involving developers more? Along the lines of "AdRadar needs to evolve towards helping find out during regular web browsing interesting situations for which it is worth investigating further the data angle (possibly through SARs)"

Would you send us an invitation? You have the fullest agenda of all

**Comment by pdehaye** [comment]
done

**Comment by foucault-dumas** [comment]
> sellers.guide just added a ads.txt cleaner on their page we just saw an interesting webinar by them with @mvidonne. She summarized it [here](https://drive.infomaniak.com/app/office/193995/92353).
> 
> I don't know how to answer @ffsinger's question, the idea is to use ads.txt and sellers.json to add live knowledge on the knowledge harvested on adds by adradar. For example display (and explain) the links between the intermediary identify as the winning bidder for an add and other actors of the adtech ecosystem. Can ads.txt and sellers.json (and sellers.guide) be used to do so?
> 
> Edit: [sellers.guide's slides](https://drive.infomaniak.com/app/share/193995/fcf9ade1-c1c6-4f6a-919e-fab22d065c8a)

@mvidonne also found this [very interesting tool](https://pagexray.fouanalytics.com), which is like the [Markup tool](https://themarkup.org/blacklight) but more complex (and uglier)

**Comment by foucault-dumas** [comment]
Isn't that something we should dig in? https://github.com/InteractiveAdvertisingBureau/openrtb/blob/master/supplychainobject.md
